### ISSUE TITLE
when use gradient_accumulation, optimizer.zero_grad() must be put to the end of a training iteration

### ISSUE THREAD

**Issue by baofff** [huggingface/accelerate#1299]
When use gradient_accumulation, optimizer.zero_grad() must be put to the end of a training iteration. 

The official demo https://huggingface.co/docs/accelerate/v0.18.0/en/usage_guides/explore puts optimizer.zero_grad() to the beginning of a training iteration and will discard previous accumulated gradient.
